File: windows/DracPaste.Protocolo/Sesion/Handshake.cs
using DracPaste.Protocolo.Mensajes;
using DracPaste.Protocolo.Red;
using DracPaste.Protocolo.Seguridad;

namespace DracPaste.Protocolo.Sesion;

/// <summary>
/// El handshake de sesión (<c>docs/protocol.md</c> §4).
///
/// Está separado del socket a propósito: recibe un <see cref="Stream"/> cualquiera, así
/// que se puede probar entero sobre loopback, sin móvil ni WiFi.
///
/// Al terminar, los dos extremos saben que el otro posee la clave de par, y cada uno
/// tiene su pareja de sobres cifrados con los contadores a cero.
/// </summary>
public static class Handshake
{
    /// <summary>
    /// Lado del PC: espera el <c>HELLO</c> del móvil y responde.
    /// </summary>
    /// <param name="buscarClavePar">
    /// Devuelve la clave de par de ese <c>device_id</c>, o <c>null</c> si no está
    /// emparejado. Un móvil desconocido no recibe respuesta: no se le confirma siquiera
    /// que este PC exista.
    /// </param>
    public static async Task<SesionEstablecida> AceptarAsync(
        Stream flujo,
        string miDeviceId,
        Func<string, byte[]?> buscarClavePar,
        CancellationToken ct = default)
    {
        using var limite = CancellationTokenSource.CreateLinkedTokenSource(ct);
        limite.CancelAfter(Protocolo.TimeoutHandshakeMs);
        var token = limite.Token;

        if (await Framing.LeerAsync(flujo, token).ConfigureAwait(false) is not { } bytesHello ||
            CodecMensajes.Decodificar(bytesHello) is not Hello hello)
        {
            throw new ProtocoloException("Se esperaba un HELLO para abrir la sesión");
        }

        if (hello.V != Protocolo.Version)
        {
            throw new ProtocoloException($"Versión de protocolo no soportada: {hello.V}");
        }

        var clavePar = buscarClavePar(hello.DeviceId)
            ?? throw new ProtocoloException($"El dispositivo {hello.DeviceId} no está emparejado");

        var retoMovil = DecodificarReto(hello.Challenge, "del móvil");
        var retoPc = Cripto.Aleatorio(Cripto.TamReto);

        await Framing.EscribirAsync(
            flujo,
            CodecMensajes.Codificar(new ServerHello
            {
                DeviceId = miDeviceId,
                Challenge = Convert.ToBase64String(retoPc),
            }),
            token).ConfigureAwait(false);

        var claves = Derivacion.ClavesDeSesion(clavePar, retoMovil, retoPc);
        var entrante = new SobreCifrado(claves.ParaRecibir(soyElMovil: false));
        var saliente = new SobreCifrado(claves.ParaEnviar(soyElMovil: false));

        // Que este AUTH se descifre ya demuestra que el otro extremo tiene la clave de
        // par; el eco demuestra además que responde a esta sesión y no a una grabada.
        var auth = CodecMensajes.Decodificar(
            entrante.Abrir(await Framing.LeerAsync(flujo, token).ConfigureAwait(false)));

        if (auth is not Auth autenticacion)
        {
            throw new ProtocoloException($"Se esperaba un AUTH y llegó {auth.T}");
        }

        VerificarEco(autenticacion.Echo, retoPc);

        await Framing.EscribirAsync(
            flujo,
            saliente.Sellar(CodecMensajes.Codificar(new AuthOk { Echo = Convert.ToBase64String(retoMovil) })),
            token).ConfigureAwait(false);

        return new SesionEstablecida(hello.DeviceId, entrante, saliente);
    }

    /// <summary>
    /// Lado del móvil: abre la sesión contra un PC ya emparejado.
    ///
    /// Existe también en C# —aunque el cliente real sea Android— porque permite probar
    /// el handshake completo de punta a punta sobre loopback, y porque un futuro cliente
    /// de escritorio lo usaría tal cual.
    /// </summary>
    /// <param name="deviceIdEsperado">
    /// El <c>device_id</c> del PC activo. Si contesta otro, se corta: es justo lo que
    /// haría un impostor de la red anunciando el mismo servicio mDNS.
    /// </param>
    public static async Task<SesionEstablecida> IniciarAsync(
        Stream flujo,
        string miDeviceId,
        string deviceIdEsperado,
        byte[] clavePar,
        CancellationToken ct = default)
    {
        using var limite = CancellationTokenSource.CreateLinkedTokenSource(ct);
        limite.CancelAfter(Protocolo.TimeoutHandshakeMs);
        var token = limite.Token;

        var retoMovil = Cripto.Aleatorio(Cripto.TamReto);
        await Framing.EscribirAsync(
            flujo,
            CodecMensajes.Codificar(new Hello
            {
                DeviceId = miDeviceId,
                Challenge = Convert.ToBase64String(retoMovil),
            }),
            token).ConfigureAwait(false);

        if (CodecMensajes.Decodificar(await Framing.LeerAsync(flujo, token).ConfigureAwait(false))
            is not ServerHello serverHello)
        {
            throw new ProtocoloException("Se esperaba un SERVER_HELLO");
        }

        if (serverHello.V != Protocolo.Version)
        {
            throw new ProtocoloException($"Versión de protocolo no soportada: {serverHello.V}");
        }

        if (serverHello.DeviceId != deviceIdEsperado)
        {
            throw new ProtocoloException(
                $"Contestó {serverHello.DeviceId} y se esperaba {deviceIdEsperado}");
        }

        var retoPc = DecodificarReto(serverHello.Challenge, "del PC");
        var claves = Derivacion.ClavesDeSesion(clavePar, retoMovil, retoPc);
        var entrante = new SobreCifrado(claves.ParaRecibir(soyElMovil: true));
        var saliente = new SobreCifrado(claves.ParaEnviar(soyElMovil: true));

        await Framing.EscribirAsync(
            flujo,
            saliente.Sellar(CodecMensajes.Codificar(new Auth { Echo = Convert.ToBase64String(retoPc) })),
            token).ConfigureAwait(false);

        var respuesta = CodecMensajes.Decodificar(
            entrante.Abrir(await Framing.LeerAsync(flujo, token).ConfigureAwait(false)));

        if (respuesta is not AuthOk authOk)
        {
            throw new ProtocoloException($"Se esperaba un AUTH_OK y llegó {respuesta.T}");
        }

        VerificarEco(authOk.Echo, retoMovil);

        return new SesionEstablecida(serverHello.DeviceId, entrante, saliente);
    }

    private static byte[] DecodificarReto(string base64, string dueno)
    {
        byte[] reto;
        try
        {
            reto = Convert.FromBase64String(base64);
        }
        catch (FormatException e)
        {
            throw new ProtocoloException($"El reto {dueno} no es base64 válido", e);
        }

        if (reto.Length != Cripto.TamReto)
        {
            throw new ProtocoloException(
                $"El reto {dueno} tiene {reto.Length} bytes y debe tener {Cripto.TamReto}");
        }

        return reto;
    }

    private static void VerificarEco(string ecoBase64, byte[] esperado)
    {
        byte[] eco;
        try
        {
            eco = Convert.FromBase64String(ecoBase64);
        }
        catch (FormatException e)
        {
            throw new ProtocoloException("El eco del reto no es base64 válido", e);
        }

        if (!Cripto.IgualesEnTiempoConstante(eco, esperado))
        {
            throw new ProtocoloException("El eco del reto no coincide");
        }
    }
}

/// <summary>Una sesión autenticada, con sus dos sobres listos.</summary>
public sealed class SesionEstablecida
{
    public SesionEstablecida(string deviceIdRemoto, SobreCifrado entrante, SobreCifrado saliente)
    {
        DeviceIdRemoto = deviceIdRemoto;
        Entrante = entrante;
        Saliente = saliente;
    }

    public string DeviceIdRemoto { get; }

    public SobreCifrado Entrante { get; }

    public SobreCifrado Saliente { get; }

    public void Limpiar()
    {
        Entrante.Limpiar();
        Saliente.Limpiar();
    }
}

File: android/protocolo/src/main/kotlin/com/marcmayol/dracpaste/protocolo/sesion/Handshake.kt
package com.marcmayol.dracpaste.protocolo.sesion

import com.marcmayol.dracpaste.protocolo.Protocolo
import com.marcmayol.dracpaste.protocolo.cripto.Cripto
import com.marcmayol.dracpaste.protocolo.cripto.Derivacion
import com.marcmayol.dracpaste.protocolo.mensajes.Auth
import com.marcmayol.dracpaste.protocolo.mensajes.AuthOk
import com.marcmayol.dracpaste.protocolo.mensajes.CodecMensajes
import com.marcmayol.dracpaste.protocolo.mensajes.Hello
import com.marcmayol.dracpaste.protocolo.mensajes.ServerHello
import com.marcmayol.dracpaste.protocolo.red.Framing
import com.marcmayol.dracpaste.protocolo.red.ProtocoloException
import com.marcmayol.dracpaste.protocolo.red.SobreCifrado
import java.io.InputStream
import java.io.OutputStream
import java.util.Base64

/**
 * El handshake de sesión (`docs/protocol.md` §4).
 *
 * Está separado del socket a propósito: recibe flujos cualesquiera, así que se puede
 * probar entero sobre loopback, sin móvil ni WiFi.
 *
 * Al terminar, los dos extremos saben que el otro posee la clave de par, y cada uno
 * tiene su pareja de sobres cifrados con los contadores a cero.
 *
 * Las operaciones son bloqueantes: en Android se llaman desde `Dispatchers.IO`.
 */
object Handshake {

    /**
     * Lado del móvil: abre la sesión contra un PC ya emparejado.
     *
     * @param deviceIdEsperado `device_id` del PC activo. Si contesta otro, se corta: es
     *   justo lo que haría un impostor de la red anunciando el mismo servicio mDNS.
     */
    fun iniciar(
        entrada: InputStream,
        salida: OutputStream,
        miDeviceId: String,
        deviceIdEsperado: String,
        clavePar: ByteArray,
    ): SesionEstablecida {
        val retoMovil = Cripto.aleatorio(Cripto.TAM_RETO)
        Framing.escribir(
            salida,
            CodecMensajes.codificar(
                Hello(deviceId = miDeviceId, challenge = Base64.getEncoder().encodeToString(retoMovil)),
            ),
        )
        salida.flush()

        val serverHello = CodecMensajes.decodificar(Framing.leer(entrada)) as? ServerHello
            ?: throw ProtocoloException("Se esperaba un SERVER_HELLO")

        if (serverHello.v != Protocolo.VERSION) {
            throw ProtocoloException("Versión de protocolo no soportada: ${serverHello.v}")
        }
        if (serverHello.deviceId != deviceIdEsperado) {
            throw ProtocoloException(
                "Contestó ${serverHello.deviceId} y se esperaba $deviceIdEsperado",
            )
        }

        val retoPc = decodificarReto(serverHello.challenge, "del PC")
        val claves = Derivacion.clavesDeSesion(clavePar, retoMovil, retoPc)
        val entranteCifrado = SobreCifrado(claves.paraRecibir(soyElMovil = true))
        val salienteCifrado = SobreCifrado(claves.paraEnviar(soyElMovil = true))

        Framing.escribir(
            salida,
            salienteCifrado.sellar(
                CodecMensajes.codificar(Auth(echo = Base64.getEncoder().encodeToString(retoPc))),
            ),
        )
        salida.flush()

        val respuesta = CodecMensajes.decodificar(entranteCifrado.abrir(Framing.leer(entrada)))
        val authOk = respuesta as? AuthOk
            ?: throw ProtocoloException("Se esperaba un AUTH_OK y llegó ${respuesta.t}")

        verificarEco(authOk.echo, retoMovil)

        return SesionEstablecida(serverHello.deviceId, entranteCifrado, salienteCifrado)
    }

    /**
     * Lado del PC. En Android no se usa —el móvil siempre es el cliente— pero existe
     * para poder probar el handshake completo de punta a punta sobre loopback.
     *
     * @param buscarClavePar devuelve la clave de par de ese `device_id`, o `null` si no
     *   está emparejado.
     */
    fun aceptar(
        entrada: InputStream,
        salida: OutputStream,
        miDeviceId: String,
        buscarClavePar: (String) -> ByteArray?,
    ): SesionEstablecida {
        val hello = CodecMensajes.decodificar(Framing.leer(entrada)) as? Hello
            ?: throw ProtocoloException("Se esperaba un HELLO para abrir la sesión")

        if (hello.v != Protocolo.VERSION) {
            throw ProtocoloException("Versión de protocolo no soportada: ${hello.v}")
        }

        val clavePar = buscarClavePar(hello.deviceId)
            ?: throw ProtocoloException("El dispositivo ${hello.deviceId} no está emparejado")

        val retoMovil = decodificarReto(hello.challenge, "del móvil")
        val retoPc = Cripto.aleatorio(Cripto.TAM_RETO)

        Framing.escribir(
            salida,
            CodecMensajes.codificar(
                ServerHello(deviceId = miDeviceId, challenge = Base64.getEncoder().encodeToString(retoPc)),
            ),
        )
        salida.flush()

        val claves = Derivacion.clavesDeSesion(clavePar, retoMovil, retoPc)
        val entranteCifrado = SobreCifrado(claves.paraRecibir(soyElMovil = false))
        val salienteCifrado = SobreCifrado(claves.paraEnviar(soyElMovil = false))

        val mensaje = CodecMensajes.decodificar(entranteCifrado.abrir(Framing.leer(entrada)))
        val auth = mensaje as? Auth
            ?: throw ProtocoloException("Se esperaba un AUTH y llegó ${mensaje.t}")

        verificarEco(auth.echo, retoPc)

        Framing.escribir(
            salida,
            salienteCifrado.sellar(
                CodecMensajes.codificar(AuthOk(echo = Base64.getEncoder().encodeToString(retoMovil))),
            ),
        )
        salida.flush()

        return SesionEstablecida(hello.deviceId, entranteCifrado, salienteCifrado)
    }

    private fun decodificarReto(base64: String, dueno: String): ByteArray {
        val reto = try {
            Base64.getDecoder().decode(base64)
        } catch (e: IllegalArgumentException) {
            throw ProtocoloException("El reto $dueno no es base64 válido", e)
        }
        if (reto.size != Cripto.TAM_RETO) {
            throw ProtocoloException(
                "El reto $dueno tiene ${reto.size} bytes y debe tener ${Cripto.TAM_RETO}",
            )
        }
        return reto
    }

    private fun verificarEco(ecoBase64: String, esperado: ByteArray) {
        val eco = try {
            Base64.getDecoder().decode(ecoBase64)
        } catch (e: IllegalArgumentException) {
            throw ProtocoloException("El eco del reto no es base64 válido", e)
        }
        if (!Cripto.igualesEnTiempoConstante(eco, esperado)) {
            throw ProtocoloException("El eco del reto no coincide")
        }
    }
}

/** Una sesión autenticada, con sus dos sobres listos. */
class SesionEstablecida(
    val deviceIdRemoto: String,
    val entrante: SobreCifrado,
    val saliente: SobreCifrado,
) {
    fun limpiar() {
        entrante.limpiar()
        saliente.limpiar()
    }
}
